exporter/exporterhelper/tracehelper.go

```diff
@@ -116,3 +129,104 @@ func pushTraceDataWithSpan(next PushTraceData, spanName string) PushTraceData {
 		return droppedSpans, err
 	}
 }
+
+type otlpTraceExporter struct {
+	exporterFullName string
+	pushTraceData    PushOTLPTraceData
+	shutdown         Shutdown
+}
+
+var _ exporter.OTLPTraceExporter = (*otlpTraceExporter)(nil)
+
+func (te *otlpTraceExporter) Start(host component.Host) error {
+	return nil
+}
+
+func (te *otlpTraceExporter) ConsumeOTLPTrace(
+	ctx context.Context,
+	td consumerdata.OTLPTraceData,
+) error {
+	exporterCtx := observability.ContextWithExporterName(ctx, te.exporterFullName)
+	_, err := te.pushTraceData(exporterCtx, td)
+	return err
+}
+
+// Shutdown stops the exporter and is invoked during shutdown.
+func (te *otlpTraceExporter) Shutdown() error {
+	return te.shutdown()
+}
+
+// NewOTLPTraceExporter creates an OTLPTraceExporter that can record metrics and can wrap
+// every request with a Span.
+func NewOTLPTraceExporter(
+	config configmodels.Exporter,
+	pushTraceData PushOTLPTraceData,
+	options ...ExporterOption,
+) (exporter.OTLPTraceExporter, error) {
+
+	if config == nil {
+		return nil, errNilConfig
+	}
+
+	if pushTraceData == nil {
+		return nil, errNilPushTraceData
+	}
+
+	opts := newExporterOptions(options...)
+	if opts.recordMetrics {
+		pushTraceData = pushOTLPTraceDataWithMetrics(pushTraceData)
+	}
+
+	if opts.recordTrace {
+		spanName := config.Name() + spanNameSuffix
+		pushTraceData = pushOTLPTraceDataWithSpan(pushTraceData, spanName)
+	}
+
+	// The default shutdown function does nothing.
+	if opts.shutdown == nil {
+		opts.shutdown = func() error {
+			return nil
+		}
+	}
+
+	return &otlpTraceExporter{
+		exporterFullName: config.Name(),
+		pushTraceData:    pushTraceData,
+		shutdown:         opts.shutdown,
+	}, nil
+}
+
+func pushOTLPTraceDataWithMetrics(next PushOTLPTraceData) PushOTLPTraceData {
```

Comment:
nit: may be rename these to something like `wrapMeterics/Traces` and same for traces? At first it looked like the function was actually pushing data.

Reply:
On second thought I think it's fine as is. This does follow the common pattern for naming wrapper like things. It's just that the function type name is a verb and threw me off. 

May be rename the type to something like `otlpTraceDataPusher` ? I think that would be a more fitting name for a type/interface.

Reply:
`PushTraceData` is already an exported name, I don't want to rename it and risk breaking others code. Because of that and for naming consistency I think it is best to use `PushOTLPTraceData` name.

Not sure if worth renaming non-exported function names to something like `traceDataWithMetricsPusher` or just `traceDataWithMetrics`. I think using "with" in the name is common in our codebase to be clear enough, so "Pusher" suffix may not be even necessary (e.g. option modifier function generators are usually named "withSomething").

Thoughts?

Reply:
I guess we can leave things as are for consistency.

Reply:
But here is another idea: Since we are defining a new type, we can add "modifier" methods to that type. Something like:

func (p PushOTLPTraceData) withSpan(spanName string) PushOTLPTraceData {
     // return wrapped func that calls `p()` internally
}

Reply:
I am not sure I like the suggestion.

This implies that all "withNNNN" function generators should all be members of PushOTLPTraceData (otherwise it would be very inconsistent if some were members and others were free-standing functions).

However, that automatically disallows creating function generators outside the package were PushOTLPTraceData is defined.

Reply:
Actually I don't know why PushTraceData is exported at all.

I searched contrib and our custom build, it is not used anywhere outside its package. It does not need to be exported. At this point I am inclined to ignore the fact that it was exported and just rename/re-work it as we want. I don't this will break anyones code (and if it does it is trivial to fix).

I would rename PushTraceData to traceDataPusher and pushTraceDataWithMetrics simply to withMetrics, etc (since these are private there is no need for long names).

I could then even do what you suggested and make withMetrics, etc members of traceDataPusher. My earlier argument about the need to define them outside this package is no longer valid.

Thoughts? (cc @pjanotti )

Reply:
I don't think implies everything must be a method outside this package. External packages are free to use the contract however they wish. But it doesn't matter that much. I personally do like the `.withX` method but it's a minor thing and we can keep things as they are if you want. Not worth delaying the PR anymore.

Reply:
(race condition in comments. was meant as a reply to earlier comment :) )

Reply:
OK. So you like my last suggestion?

Reply:
Yes.

Reply:
@pjanotti what do you think?

Reply:
Refactored, please have another look.